### PR TITLE
Use different ways to set ciphers for TLS 1.3

### PR DESCRIPTION
**Pull request by ghost**
Openssl have a different way to set ciphers for TLS 1.3. If you using ssl_ciphers with only TLS 1.3 nginx will fail to start. You need to use "ssl_conf_command Ciphersuites" command in nginx.

I implement it that way that I check in the template if TLS version is set to 1.3 if so use the new syntax.

Links:
https://forum.nginx.org/read.php?11,287698
https://github.com/mozilla/ssl-config-generator/issues/124
https://wiki.openssl.org/index.php/TLS1.3

### CODE REVIEW

**Comment by chr4** [comment]
Good catch. I'm not convinced yet to use the same variable for both cipher suites, though. 

TLSv13 and TLSv12 might be used alongside. The current config doesn't allow for TLSv13 cipher suite specification (not critical, as they are by default quite good). But currently, it's not possible to specify `ssl_conf_command Ciphersuites` for a mixed configuration - and the existing `ssl_ciphers` option will then be automatically mapped to TLSv13 once `ssl_protocols` is set to only use TLSv13.

I'd vote for a cleaner approach on using a second option, either `ssl_conf_command` directly, or abstracting it as `tlsv13_ciphers` (I think the former would be a bit clearer/cleaner), that can be used to specify TLSv13 ciphers if set.

I'd still set `ssl_ciphers` when TLSv13 only is enabled, just so it fails and people get feedback about the invalid configuration. Ignoring it silently in TLSv13 mode might lead the caller to believe they set it correctly, but it's silently ignored.

**Comment by dgo-** [comment]
Make sense!

I would go for the approve add an extra variable(pillar) `ssl_conf_command`.  (Note: [ssl_conf_command](https://www.openssl.org/docs/man1.1.1/man3/SSL_CONF_cmd.html), can used to configure more then ciphers)

I would set default the `ssl_conf_command` variable to an empty list. I would choose a list here, because you can you `ssl_conf_command` multiple times. See [example from nginx docs](http://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_conf_command):
```
ssl_conf_command Options PrioritizeChaCha;
ssl_conf_command Ciphersuites TLS_CHACHA20_POLY1305_SHA256;
```
I would default to an empty list, because the nginx default is also not set. In the template  would do a for loop over the list, so we only set it if something is in the list. 

As you suggested I would always set the `ssl_ciphers` variables. If `ssl_conf_command Ciphersuites` is set correctly you can also set the `ssl_ciphers` without running in a problem or lowering the security on TLS 1.3.  

What do you think about the approve?

File: pillar.example [hunk 1]
```diff
@@ -6,6 +6,7 @@ nginx:
   # This values can be overwritten in each server context individually.
   ssl_protocols: 'TLSv1.2 TLSv1.3'
   ssl_ciphers: 'ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256' # noqa: 204
+  ssl_conf_command: []
```
Comment (chr4) [inline pillar.example]: I think you should add a "real" example on how to use this, but mention the default is empty (maybe in a seperate paragraph), then test for that change to actually be in the file in the test below.

**Comment by chr4** [comment]
Yes, that sounds like a good plan! I think you can use the example you provided above in `pillar.example`, as it automatically verifies that multiple arguments are working.

File: pillar.example [hunk 2]
```diff
@@ -6,6 +6,14 @@ nginx:
   # This values can be overwritten in each server context individually.
   ssl_protocols: 'TLSv1.2 TLSv1.3'
   ssl_ciphers: 'ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256' # noqa: 204
+  # ssl_conf_command need to be a list, because it can set multiple times in nginx.conf.
+  # Each sting in the list needs to hold all arguments that needs to go to a single line.
+  # See example below.
+  # By default ssl_conf_command is set to an empty list, which result in no entry in the nginx.conf
+  # ssl_conf_command: []
+  # example with values:
+  ssl_conf_command: [ "Options PrioritizeChaCha", "Ciphersuites TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384"]
```
Comment (chr4) [inline pillar.example]:
```suggestion

# ssl_conf_command can be specified multiple times, therefore it needs to be a list.
# Each string needs to hold all arguments for ssl_conf_command, like in the example below.
# An empty list ([], the default) results in no entries added to nginx.conf
ssl_conf_command: [ 'Options PrioritizeChaCha', 'Ciphersuites TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384']

Comment (chr4) [reply pillar.example]: reduced the comment a little, some typos, also used single quotes for consistency.

Comment (chr4) [reply pillar.example]: wdyt?

Comment (dgo-) [reply pillar.example]: looks good.

File: pillar.example [hunk 3]
```diff
@@ -7,6 +7,12 @@ nginx:
   ssl_protocols: 'TLSv1.2 TLSv1.3'
   ssl_ciphers: 'ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256' # noqa: 204
 
+# ssl_conf_command can be specified multiple times, therefore it needs to be a list.
+# Each string needs to hold all arguments for ssl_conf_command, like in the example below.
+# An empty list ([], the default) results in no entries added to nginx.conf
+ssl_conf_command: [ 'Options PrioritizeChaCha', 'Ciphersuites TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384']
+
+
```
Comment (chr4) [inline pillar.example]:
```suggestion
  # ssl_conf_command can be specified multiple times, therefore it needs to be a list.
  # Each string needs to hold all arguments for ssl_conf_command, like in the example below.
  # An empty list ([], the default) results in no entries added to nginx.conf
  ssl_conf_command: [ 'Options PrioritizeChaCha', 'Ciphersuites TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384']
```

Also, seems like Github suggestions can't remove empty lines? I wanted only one.

Comment (chr4) [reply pillar.example]: messed up indention

File: pillar.example [hunk 4]
```diff
@@ -7,6 +7,12 @@ nginx:
   ssl_protocols: 'TLSv1.2 TLSv1.3'
   ssl_ciphers: 'ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256' # noqa: 204
 
+# ssl_conf_command can be specified multiple times, therefore it needs to be a list.
+# Each string needs to hold all arguments for ssl_conf_command, like in the example below.
+# An empty list ([], the default) results in no entries added to nginx.conf
+ssl_conf_command: [ 'Options PrioritizeChaCha', 'Ciphersuites TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384']
+
```
Comment (chr4) [inline pillar.example]:
```suggestion
```

**Comment by chr4** [comment]
Perfect, thanks!

**Pull request merged**